source/CollisionSet.cpp

```diff
@@ -164,25 +167,51 @@ void CollisionSet::Finish()
 
 
 
-// Get the first object that collides with the given projectile. If a
-// "closest hit" value is given, update that value.
-Body *CollisionSet::Line(const Projectile &projectile, double *closestHit) const
+// Get the first (closest) collision for the given projectile.
+Collision CollisionSet::Line(const Projectile &projectile) const
 {
 	// What objects the projectile hits depends on its government.
 	const Government *pGov = projectile.GetGovernment();
 
 	// Convert the start and end coordinates to integers.
 	Point from = projectile.Position();
 	Point to = from + projectile.Velocity();
-	return Line(from, to, closestHit, pGov, projectile.Target());
+	return Line(from, to, pGov, projectile.Target());
 }
 
 
 
-// Check for collisions with a line, which may be a projectile's current
+// Get all collisions for the given projectile. Collisions are not necessarily
+// sorted by distance.
+const vector<Collision> &CollisionSet::LineAll(const Projectile &projectile) const
+{
+	// What objects the projectile hits depends on its government.
+	const Government *pGov = projectile.GetGovernment();
+
+	// Convert the start and end coordinates to integers.
+	Point from = projectile.Position();
+	Point to = from + projectile.Velocity();
+	// If this projectile can only hit a single target anyway then just get the closest
+	// hit.
+	return LineAll(from, to, pGov, projectile.Target(), projectile.HitsRemaining() == 1);
+}
+
+
+
+// Get the first (closest) collision along a line, which may be a projectile's current
 // position or its entire expected trajectory (for the auto-firing AI).
-Body *CollisionSet::Line(const Point &from, const Point &to, double *closestHit,
-		const Government *pGov, const Body *target) const
+Collision CollisionSet::Line(const Point &from, const Point &to, const Government *pGov,
+		const Body *target) const
+{
+	return LineAll(from, to, pGov, target, false)[0];
+}
+
+
+
+// Get all collisions along a line, given that the "all" boolean is true.
+// Otherwise, return the first (closest) collision.
+const vector<Collision> &CollisionSet::LineAll(const Point &from, const Point &to,
+		const Government *pGov, const Body *target, bool all) const
```

Comment:
Couldn't we also pass an amount? So that we get only the collisions we need?

Reply:
We could stop after N collisions, but how would we know those are the closest N collisions?

Reply:
true I forgot